### PR TITLE
[ENH] EnbPI based forecaster

### PR DESCRIPTION
**Pull request by benHeid**
#### Reference Issues/PRs
Towards #6446, uses the aws-fortuna based approach together with tsbootstrap bootstraps

#### What does this implement/fix? Explain your changes.
Interfaces the aws-fortuna EnbPI implementation

#### Does your contribution introduce a new dependency? If yes, which one?
Yes aws-fortuna as well as third party library to use them (Jax==0.28.1)

#### Did you add any tests for the change?

get_test_params

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

File: sktime/forecasting/compose/_enbpi_forecaster.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+            )
+            self.preds.append(prediction)
+
+        return self
+
+    def _predict(self, X, fh=None):
+        # Calculate Prediction Intervals using Bootstrap Samples
+
+        preds = []
+        for forecaster in self.forecasters:
```
Comment (VascoSch92) [inline sktime/forecasting/compose/_enbpi_forecaster.py:129]: You could use a list compression here and it should be a little faster :)

**Comment by benHeid** [comment]
@fkiraly should be ready for review :)

**Comment by benHeid** [comment]
Before merging check the dependencies..

**Comment by benHeid** [comment]
For the EnbPI wrapper, we using just a single method of the EnbPI class in the aws-fortuna package. Unfortunately this package has a lot of soft dependencies, including JAX. Furthermore, currently, the package is not easily installable due to dependency conflicts: https://github.com/awslabs/fortuna/issues/194

Thus, I wonder if we could copy paste the method conformal_intervals method from the EnbPI class of the aws-fortuna package and adapt it slightly (adaption would get rid off the checks that aren't required anymore since it is internal and replacement of jax.numpy by numpy directly. If copying and modifiying we need to include the original license and the notice statement. Not sure how we should do this: https://github.com/awslabs/fortuna?tab=Apache-2.0-1-ov-file#readme and https://github.com/awslabs/fortuna/blob/main/NOTICE

@fkiraly any ideas?

**Comment by fkiraly** [comment]
If it is a very small part only, it could go in a private submodule with a license file. We could also give the user the option to use `fortuna` if they manage to get it installed, via a hyperparameter, but that´s perhaps overkill.

File: sktime/forecasting/compose/_enbpi_forecaster.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        bs_ts_index = list(
+            self.bootstrap_transformer_.bootstrap(y, test_ratio=0, return_indices=True)
+        )
+        self.indexes = np.stack(list(map(lambda x: x[1], bs_ts_index)))
+        bootstapped_ts = list(map(lambda x: x[0], bs_ts_index))
+
+        self.forecasters = []
+        self.preds = []
+        # Fit Models per Bootstrap Sample
+        for bs_ts in bootstapped_ts:
```
Comment (fkiraly) [inline sktime/forecasting/compose/_enbpi_forecaster.py:119]: probably a typo - "bootstapped"

File: sktime/forecasting/compose/_enbpi_forecaster.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+            bs_df = pd.DataFrame(bs_ts, index=y.index)
+            forecaster = clone(self.forecaster_)
+            forecaster.fit(y=bs_df, fh=fh, X=X)
+            self.forecasters.append(forecaster)
+            prediction = forecaster.predict(fh=y.index, X=X)
```
Comment (fkiraly) [inline sktime/forecasting/compose/_enbpi_forecaster.py:124]: why are we making predictions here? It does not seem the `self.preds` is ever used again.

File: sktime/forecasting/compose/_enbpi_forecaster.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+            [self._y.columns, coverage, ["lower", "upper"]]
+        )
+        fh_absolute = fh.to_absolute(self.cutoff)
+        fh_absolute_idx = fh_absolute.to_pandas()
+        pred_int = pd.DataFrame(
+            np.concatenate(intervals, axis=1), index=fh_absolute_idx, columns=cols
+        )
+        return pred_int
+
+    def _update(self, y, X=None, update_params=True):
```
Comment (fkiraly) [inline sktime/forecasting/compose/_enbpi_forecaster.py:170]: I think this is equivalent to the default, so could be removed?

File: sktime/forecasting/compose/_enbpi_forecaster.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+                bootstrap_test_preds=np.stack(preds),
+                train_targets=train_targets,
+                error=1 - cov,
+            )
+            intervals.append(conformal_intervals)
+
+        cols = pd.MultiIndex.from_product(
+            [self._y.columns, coverage, ["lower", "upper"]]
+        )
+        fh_absolute = fh.to_absolute(self.cutoff)
```
Comment (fkiraly) [inline sktime/forecasting/compose/_enbpi_forecaster.py:163]: same as `to_absolute_index`

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks good, some minor things:

* for `sktime` estimators, the `clone` method (of the estimator) should be used, or configs will not be cloned.
* the docstring should describe better how the algorithm is applied
* I would move it to an enbpi module. `compose` is for pipelines or similar, it does have some "ensembles" in it but I think in some cases these are misnomers, in others I would not have put it there.

Non-blocking: some typos

File: sktime/forecasting/compose/_enbpi_forecaster.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+from sktime.forecasting.base import BaseForecaster
+from sktime.forecasting.naive import NaiveForecaster
+
+
+class EnbPIForecaster(BaseForecaster):
+    """
+    Ensemble Bootstrap Prediction Interval Forecaster.
+
+    The forecaster combines sktime forecasters, with tsbootratp bootsrappers
```
Comment (fkiraly) [inline sktime/forecasting/compose/_enbpi_forecaster.py:14]: another typo

File: sktime/forecasting/compose/_enbpi_forecaster.py (hunk 2 above)
Comment (benHeid) [reply sktime/forecasting/compose/_enbpi_forecaster.py:119]: Done

File: sktime/forecasting/compose/_enbpi_forecaster.py (hunk 3 above)
Comment (benHeid) [reply sktime/forecasting/compose/_enbpi_forecaster.py:124]: Done

File: sktime/forecasting/compose/_enbpi_forecaster.py (hunk 6 above)
Comment (benHeid) [reply sktime/forecasting/compose/_enbpi_forecaster.py:14]: Done

File: sktime/forecasting/compose/_enbpi_forecaster.py (hunk 4 above)
Comment (benHeid) [reply sktime/forecasting/compose/_enbpi_forecaster.py]: Done

**Comment by benHeid** [comment]
@fkiraly I addressed your changes and additionally also added a adapted version of the EnbPI implementation from aws-fortuna into the libs folder, together with the corresponding license and the NOTICE file that includes the copyright of aws-fortuna. I would appreciate a check if this is added correctly or if I should change the structure..

**Review by fkiraly — CHANGES_REQUESTED** [review]
Regarding the library:

* I would make it private, since it is not intended for public consumption
* call it `_aws_fortuna_enbpi` or similar as it is not the entire library, just a method from it
* `README` should be markdown, `md`
* I would also suggest to add to the main `README` in `libs`. I would suggest to add a new section "snippets" or similar and explain that these are private
    * while you are at it, you could also fix the "distibution" type :-)

Also worth considering: instead, new folder `_libs_private`, `libs_private`, or similar?

**Comment by benHeid** [comment]
I addressed your comments and created a new issue #6648

**Review by fkiraly — DISMISSED** [review]

File: sktime/forecasting/compose/__init__.py [hunk 7]
```diff
@@ -5,6 +5,7 @@
 __all__ = [
     "HierarchyEnsembleForecaster",
     "ColumnEnsembleForecaster",
+    "EnbPIForecaster",
```
Comment (fkiraly) [inline sktime/forecasting/compose/__init__.py:8]: left over?

Comment (benHeid) [reply sktime/forecasting/compose/__init__.py]: Yes..

**Comment by benHeid** [comment]
@fkiraly Please wait before merging, I need to check if I introduced a Bug. I will ping you if everything is fine.

**Comment by benHeid** [comment]
@fkiraly bug should be fixed now.

**Review by fkiraly — APPROVED** [review]

**Comment by wistuba** [comment]
fyi the soft conflict in fortuna was resolved in 0.2.0. I'm happy to take contributions that would remove soft dependencies to make the package easier to consume by others

**Pull request merged**